[AV-43732] Search Developer Guide for Server 7.2

sarahlwelton opened the pull request:
The long awaited PR for Search. 

Please review the content on the staging site, it's much easier than trying to review all of the raw content: 

https://docs-staging.couchbase.com/server/7.2/search/search.html

sarahlwelton committed:
[AV-43732] Initial commit to new repo of search devex

sarahlwelton committed:
[AV-43732] Creating a difference in a page for testing

sarahlwelton committed:
[AV-43732] Adding version to nav

sarahlwelton committed:
[AV-43732] Adding another change to a file for testing

sarahlwelton committed:
[AV-43732] Moved content from create-child-field-steps.adoc into create-child-field.adoc

Added content to make it clear that the file is for Server

sarahlwelton committed:
[AV-43732] Commit to move over all work from docs-server repo

sarahlwelton committed:
[AV-43732] Removed all mentions of Couchbase Capella and removed tabs attributes.

Added introductory sentences to procedures.

sarahlwelton committed:
[AV-43732] Quick pass to fix xrefs, partials, and examples

sarahlwelton committed:
[AV-43732] Removing last issue with xrefs

sarahlwelton committed:
[AV-43732] Added autocomplete topics to nav.adoc

Moved location of simple-search-ui.adoc in nav.adoc.

Did a consistency pass on all existing topics to add descriptions...

... remove instructions for creating an item from a fresh search index and just add "have a search index" to prereqs...

... and fix up other inconsistencies in text.

Rewrote all content in search-request-params.adoc and created a run-search-full-request.jsonc file for the examples

sarahlwelton committed:
[AV-43732] Added [abstract] attribute back to all pages.

Deleted code samples that were taken from Couchbase blog posts in favour of other samples.

Reorganized files for autocomplete, including changing order in nav.adoc

Wrote procedures for Web Console and REST API autocomplete

sarahlwelton committed:
[AV-43732] Shortened titles for index customization topics.

Fixed typo in create-custom-token-filter.adoc

Added table from search-capella branch to create-custom-tokenizer.adoc and corrected typo

Corrected typo in create-custom-wordlist.adoc

Added descriptions for some missing token filters to default-token-filters-reference.adoc

Added descriptions for all wordlists to default-wordlists-reference.adoc

Changed name of supported-languages.adoc to quick-index-supported-languages.adoc

Added content to description and added links to run-searches.adoc

Added and edited steps in search-query-auto-complete-code.adoc

Edited steps in search-query-auto-complete-ui.adoc

Changed filename of search-query-autocomplete.adoc for consistency. Added content.

Added JSON to title of search-request-params.adoc

Changed title of search.adoc and changed description

Added link to JSON properties for search requests to simple-search-rest-api.adoc

Added step to simple-search-ui.adoc

Fixed typo and changed depth of auto-complete topics in nav.adoc.

sarahlwelton committed:
[AV-43732] Final polish and review.

Lots of clarity and consistency edits.

Moved json paylods for search index into separate files for a better user experience, rather than having two indexes in one file.

Made default character filters and default tokenizers topics instead of leaving them in customize-index.adoc

Put links for SDKs into a partial to make management easier.

Rearranged and fixed some links in the nav.

Changed variables in all code samples.

sarahlwelton committed:
[AV-43732] Removing version from nav.adoc strings

sarahlwelton committed:
[AV-43732] Escaping missed {}

sarahlwelton committed:
[AV-43732] Adding version coordinate to try and fix broken links on staging

sarahlwelton committed:
Merge branch 'release/7.2' into search-server

osfameron commented on modules/search/examples/autocomplete-sample.cs:
```diff
@@ -0,0 +1,38 @@
+
+await using var cluster = await Cluster.ConnectAsync(new ClusterOptions
+{
+    ConnectionString = "CB_HOSTNAME", //
```
trailing empty comment

osfameron commented on modules/search/examples/autocomplete-sample.js:
```diff
@@ -0,0 +1,32 @@
+  const inputStr = "Beaufort Hotel"
+  for (let i = 2; i <= 8; i++) {
+    var testStr = inputStr.substring(0, i);
+    console.log("Input <" + testStr + ">, len: " + testStr.length)
```
should probably use the backtick syntax to neaten it up

sarahlwelton replied:
Can you show me what you mean? I didn't write this code sample.

osfameron replied:
```
`Input <${testStr}>, len: ${testStr.length}`
```

(This is a niggle, don't worry about it, it's just I noticed some of the other language samples did it with equivalent mechanisms, and I wondered why not in JS)

sarahlwelton replied:
If it's fine as is, then maybe leave it? lol

osfameron replied:
Yup!

osfameron commented on modules/search/examples/create-search-index-header.sh:
```diff
@@ -0,0 +1,3 @@
+curl -s -XPUT -H "Content-Type: application/json" \
+-u $CB_USERNAME:$CB_PASSWORD http://$CB_HOSTNAME:8094/api/index/landmark-content-index 
```
indentation

sarahlwelton replied:
Do you mean an indent needs to be added or that it needs to be removed?

osfameron replied:
So, I would write:

```
curl -s .... \
  -u ... \
  -whatever
```

e.g. push the flags that are on a new line so they are more clearly *contained* within the same `curl` command.
Bash doesn't really have a style-guide per se, but e.g. https://lug.fh-swf.de/vim/vim-bash/StyleGuideShell.en.pdf section 7 does seem to agree with indenting here.

I've never seen anyone put options after a continuation at column 0, but it seems to happen in a few of our examples here...
My guess is someone did it once, and this example is a copy/paste.

Either that, or someone has oddly configured editor settings?

sarahlwelton replied:
Will indent, this was totally a copy/paste / stolen from existing doc.

osfameron commented on modules/search/examples/create-search-index-payload.sh:
```diff
@@ -0,0 +1,64 @@
+curl -s -XPUT -H "Content-Type: application/json" \
+-u $CB_USERNAME:$CB_PASSWORD http://$CB_HOSTNAME:8094/api/index/landmark-content-index 
```
indentation

osfameron commented on modules/search/examples/run-search-full-request.jsonc:
```diff
@@ -0,0 +1,113 @@
+{
+    "query": {},
+
+    // tag::ctl[]
```
Hmm, if this kind of thing needed to be done regularly we could think about something like `include-jq::foo.json[.ctl.consistency]` to apply a jq path expression instead of having to manually tag these things 🤔

sarahlwelton replied:
Would be nice, yeah - maybe a note for future improvements?

osfameron replied:
I'll have a play... I think it should be "easy enough"...

osfameron commented on modules/search/examples/run-search-payload.sh:
```diff
@@ -0,0 +1,14 @@
+curl -XPOST -H "Content-Type: application/json" \
+-u $CB_USERNAME:$CB_PASSWORD http://$CB_HOSTNAME:8094/api/index/travel-sample.inventory.landmark-content-index/query \
```
indentation.

osfameron commented on modules/search/pages/create-child-field.adoc:
```diff
@@ -0,0 +1,122 @@
+. Click *insert child field*. 
+. In the *Field* field, enter the name of a field in your documents that you want to add or remove from the index.
++
+NOTE: The field must contain either a single value or an array. If the field contains a JSON object, xref:create-child-mapping.adoc[], instead.  
```
missing "see" before xref?

sarahlwelton replied:
The title is a verb phrase, so it appears as "If the field contains a JSON object, Create a Child Mapping, instead."

osfameron replied:
Ah, I thought that was possible, but couldn't tell from the local context.

If the upstream wording ever changes to e.g. "Creating a Child Mapping" this sentence would then be mangled, so I'd prefer to add the explicit link text in this case.

osfameron commented on modules/search/pages/create-child-field.adoc:
```diff
@@ -0,0 +1,122 @@
+To add or remove a child field from a Search index with the Couchbase Server Web Console:
+
+. Go to *Search*.
+. Click the index where you want to create a child field.
```
"Choose" might work here

osfameron commented on modules/search/pages/create-child-field.adoc:
```diff
@@ -0,0 +1,122 @@
+. Expand *Mappings*. 
+. Point to an existing mapping and click btn:[+].
+. Click *insert child field*. 
+. In the *Field* field, enter the name of a field in your documents that you want to add or remove from the index.
```
This feels like a long and very UI-procedural list, which means that every step seems to be as important, whether it's clickety clickety clickety click navigation, or thinking about *what* you want to change.

I wonder if we can chunk it a bit more high-level?

e.g. could we "Edit - Customize Index - Mappings" as if it was a menu selection?

or "Choose the existing mapping you want to edit. Click "+" then "insert child field".

sarahlwelton replied:
I'd be open to "Expand *Customize Index > Mappings*". Let's try that

osfameron commented on modules/search/pages/create-child-field.adoc:
```diff
@@ -0,0 +1,122 @@
+|====
+|Option |Description 
+
+|Searchable As a|
```
"Searchable As a" looks weird.
Move the `a|` to a new line?

(Though actually "Searchable As" still looks weird? What does it mean?)

sarahlwelton replied:
Looks weird how? Trust me, it's how it appears in the UI as an option, so that's unfortunately what we need to go with.... Very unfortunately.

osfameron replied:
OK, it probably makes sense in context. I think after seeing the `a|` I was thinking "Searchable A's?!" and found it hard to parse.

osfameron commented on modules/search/pages/create-child-field.adoc:
```diff
@@ -0,0 +1,122 @@
+
+The default value is the value set in *Field*.
+
+|Analyzer a|
```
Move `a|` to new line

osfameron commented on modules/search/pages/create-child-field.adoc:
```diff
@@ -0,0 +1,122 @@
+
+You can choose a default analyzer or a custom analyzer. 
+
+For more information about the available default analyzers, see xref:default-analyzers-reference.adoc[].
```
Can't these xrefs be inlines into line 52?

sarahlwelton replied:
For displaying nicely in the table, I liked having them on separate lines.

osfameron commented on modules/search/pages/create-child-field.adoc:
```diff
@@ -0,0 +1,122 @@
+
+For more information about how to create a custom analyzer, see xref:create-custom-analyzer.adoc[].
+
+|Index a|
```
Sorry, I really hate these `a|` notation, Asciidoc has some weird design decisions.
I'll stop mentioning every one now...

osfameron commented on modules/search/pages/create-child-field.adoc:
```diff
@@ -0,0 +1,122 @@
+
+To remove the field's values from the index, clear *Store*.
+
+|Include in _all field a|
```
Backticks for `_all`

sarahlwelton replied:
That's not how it appears in the UI. Trying to match the case and appearance of the options as they're shown in the UI.

osfameron commented on modules/search/pages/create-child-mapping.adoc:
```diff
@@ -0,0 +1,55 @@
+* You've created a type mapping. 
+For more information about how to create a type mapping on an index, see xref:create-type-mapping.adoc[].
+
+* You've logged in to the Couchbase Server Web Console. 
```
"You are" ?

sarahlwelton replied:
Parallelism with the previous lines :) That's all

osfameron replied:
Yeah, I think it's *not* quite parallel because "You have done X" in those cases is a one-and-done thing and logging in isn't.

sarahlwelton replied:
Well, logging in is a one-and-done isn't it? It's enter your username and password. It's something you have done.

osfameron commented on modules/search/pages/create-child-mapping.adoc:
```diff
@@ -0,0 +1,55 @@
+. Expand *Mappings*. 
+. Point to an existing mapping and click btn:[+].
+. Click *insert child mapping*. 
+. In the *{}* field, enter the name of a field in a document that contains a JSON object. 
```
The field is called `{}` ?
Does asciidoc not mangle this?

sarahlwelton replied:
The field is literally called {}. I didn't notice any issues in build.

osfameron commented on modules/search/pages/create-custom-date-time-parser.adoc:
```diff
@@ -0,0 +1,52 @@
++
+For more information, see the documentation about the https://pkg.go.dev/time#pkg-constants[Go Programming Language Time Package's Layout Constant^].
+. Click btn:[Add]. 
+. (Optional) To add an additional layout, repeat Steps 7-8. 
```
Hardcoding `7-8` feels fragile.
Could you just say "add additional layouts as needed" as they already know the procedure for this now?

sarahlwelton replied:
I don't know... Hmm. Let me do a think.

osfameron replied:
OK, not important (it's a bit fragile, but I'm not as worried by this as the `include::` comments, so feel free to close this comment if you don't find a much better way 👍)

osfameron commented on modules/search/pages/create-custom-token-filter.adoc:
```diff
@@ -0,0 +1,194 @@
+* <<edge-ngram,edge_ngram>>: Use a set character length to create tokens from the start or end of existing tokens.
+* <<elision,elision>>: Use a wordlist to remove elisions from input tokens.
+* <<keyword-marker,keyword_marker>>: Use a wordlist of keywords to find and create new tokens.
+* <<length,length>>: Use a set character length to filter out tokens that are too long or too short.
```
if the wording is identical you don't need the item after the comma do you?

sarahlwelton replied:
I'm just keeping it consistent with the other tags that I added

osfameron commented on modules/search/pages/create-custom-token-filter.adoc:
```diff
@@ -0,0 +1,194 @@
+[#dict-compound]
+=== Create a Custom `dict_compound` Token Filter
+
+include::partial$custom-token-filters-descriptions.adoc[lines=2..3;9..27]
```
huh? why are we hardcoding lines here? We can tag them.

sarahlwelton replied:
When I was doing local builds with the tags, the behaviour wasn't as expected. I can elaborate further if you want to discuss, but trust me, it was the only option that made sense.

osfameron replied:
Yeah, let's discuss, tagging an *adoc* should Just Work, I'd rather understand and ideally fix the issue!

osfameron commented on modules/search/pages/create-custom-token-filter.adoc:
```diff
@@ -0,0 +1,194 @@
+[#edge-ngram]
+=== Create a Custom `edge_ngram` Token Filter 
+
+include::partial$custom-token-filters-descriptions.adoc[lines=31..32;38..56]
```
Yeah, strongly dislike this lines syntax. We can do with tags/exclusions, happy to discuss!

osfameron commented on modules/search/pages/create-search-index-rest-api.adoc:
```diff
@@ -0,0 +1,55 @@
+
+To create a Search index with the REST API: 
+
+. In the command-line tool, start a `curl` command with the `XPUT` verb. 
```
"In the command-line tool, start a `curl` command" sounds odd.

osfameron replied:
As in, I don't know what it means. Oh, "the Couchbase command-line tool (CLI)" above... but no wiser.
Is that `cbshell`?
But you can run `curl` from your terminal of choice.

Or does "start" have a technical meaning in this specific CLI?

sarahlwelton replied:
I think I stole this from existing documentation, if you have a suggestion for how to improve the verbiage, I am all ears :)

osfameron replied:
OK, something like "Run the `curl` command with the `XPUT` verb." would be ok.
or just "Run `curl`" as you have an example following.

osfameron commented on modules/search/pages/customize-index.adoc:
```diff
@@ -0,0 +1,138 @@
+
+You can create the following custom filters: 
+
+* <<character-filters,>>
```
can you just omit the comma here?

sarahlwelton replied:
Does it work without the comma?

osfameron replied:
yeah, I think so. Not important either way though :)

osfameron commented on modules/search/pages/default-analyzers-reference.adoc:
```diff
@@ -0,0 +1,93 @@
+
+|inherit | If you set an analyzer to `inherit`, the Search index component inherits the xref:set-advanced-settings.adoc#default-analyzer[default analyzer set for an index]. 
+
+|ar |The `ar` analyzer uses character filters, tokenizers, and token filters designed for Arabic language searches.
```
Would this be clearer as just "Arabic"?
e.g. remove all the long duplicated text to focus on the important changing part.

sarahlwelton replied:
But it doesn't appear as "Arabic" in the UI - and I want to make it clear what the analyzer is doing. If someone isn't reading the whole page, I want them to be able to quickly CTRL + F and still get the information they need.

osfameron replied:
OK 👍

osfameron commented on modules/search/pages/default-analyzers-reference.adoc:
```diff
@@ -0,0 +1,93 @@
+
+|it |The `it` analyzer uses character filters, tokenizers, and token filters designed for Italian language searches.
+
+|keyword a|
```
might be clearer with the special analyzers all listed above, and then a list of the language codes separately?

sarahlwelton replied:
It's in alphabetical order and matches the order in the UI.

osfameron commented on modules/search/pages/default-token-filters-reference.adoc:
```diff
@@ -0,0 +1,213 @@
+|====
+|Token Filter Type |Description 
+
+|apostrophe | Removes all characters after an apostrophe (') from tokenizer results. Also removes the apostrophe.
```
add an example?

osfameron commented on modules/search/pages/default-token-filters-reference.adoc:
```diff
@@ -0,0 +1,213 @@
+|====
+|Token Filter Type |Description 
+
+|apostrophe | Removes all characters after an apostrophe (') from tokenizer results. Also removes the apostrophe.
```
That single `'` has confused github's syntax highlighter (e.g. rest of this page is italicised as it thinks it's part of the same string... 🤦 
Please double-check that the actual Asciidoc output looks fine?

sarahlwelton replied:
Verified many times, never noticed an issue.

osfameron commented on modules/search/pages/field-data-types-reference.adoc:
```diff
@@ -0,0 +1,36 @@
+* A string, as two numeric values separated by a comma. 
+* A string, as a geohash point. 
+* An array, as two floating point integers. 
+* A JSON object, as the properties `"lon"`/`"lng"` and `"lat"`.
```
"*as* the properties" ?

you don't need the string quotes for the properties do you?

osfameron commented on modules/search/pages/quick-index-field-options.adoc:
```diff
@@ -0,0 +1,69 @@
+
+To include content from the field in search results, select *Include in search results*. 
+
+To exclude the field's content from search results, clear *Include in search results*. 
```
I'm finding these oppositional pairs of sentences hard to read, as they look almost identical.

sarahlwelton replied:
That's why I tried to change the order of the words in each sentence.

osfameron replied:
OK, take it as a data-point of feedback to consider iterating on / leaving as is as you prefer 👍

osfameron commented on modules/search/pages/search-index-params.adoc:
```diff
@@ -0,0 +1,1004 @@
+
+[source,json]
+----
+include::example$simple-search-index-payload.jsonc[lines=1..14]
```
EEEK PLEASE GOD NO. I liked what you did in a different .jsonc with tags much more.

sarahlwelton replied:
Again, there is a method to my madness. It wasn't displaying properly without the line includes :(

osfameron commented on modules/search/pages/search-index-params.adoc:
```diff
@@ -0,0 +1,1004 @@
+
+Set the name of this object to the name you want for your custom character filter. 
+
+You can reference the `\{char_filter_name\}` object elsewhere in your Search index definition to use the character filter. 
```
do you need the escapes inside backticks?

sarahlwelton replied:
It wasn't displaying properly otherwise.

osfameron commented on modules/search/partials/custom-token-filters-descriptions.adoc:
```diff
@@ -0,0 +1,255 @@
+
+[source,json]
+----
+include::example$complex-search-index-payload.jsonc[lines=59..62]
```
tag

osfameron commented on modules/search/partials/nav.adoc:
```diff
@@ -0,0 +1,33 @@
+  ** xref:7.2@server:search:search.adoc[]
```
This is *very deeply nested* compared to many of our other navs - 

I found it a little hard to navigate but my use-case was "Docs Code Reviewer" 
(e.g. I wanted to see where an Asciidoc I was looking at was output in the site, and it was hard to discover it)
so this might be find for actual users?

sarahlwelton replied:
Very deeply nested for good reason, because of where it needs to appear in the nav in Search (under developer guides). Take a look when it's back on Staging and you can see how it looks not just in code :)

osfameron reviewed (changes requested):
Big PR, impressive work!
Overall amazing improvement, and most of my nitpicks inline can be improved iteratively.

Marking "Request changes" only because the includes with `[lines=...]` frighten me, as they will go out of date in a confusing and impossible-to-debug way.

That's a blunt instrument that we could use judiciously for code files that we don't control.

sarahlwelton committed:
[AV-43732] Remove duplicate modules that were causing errors.

sarahlwelton committed:
[AV-43732] Applying suggestions from code review

sarahlwelton committed:
[AV-43732] Adding indentation to code samples

sarahlwelton committed:
[AV-43732] Adding one last suggestion from code review (for now)

sarahlwelton committed:
[AV-43732] Making equivalent changes to other files from code review for consistency

osfameron committed:
Fix lines to tags

This is, admittedly a PITA, especially when VSCode "helpfully" messes
up the formatting of the JSON that you're working around, but apart
from being slightly awkward, it seems to work like this.

It feels like there are layered improvements to make this nicer.

1. use line numbers because tags are annoying

2. use tags because even so, they'll be more maintainable longer term

3. Looking at the requirements, it seems like we want to:

 * select a path from the JSON
 * elide after a certain number of lines, because the output

 So something like this syntax for an include processor might be nice?

 `include::some-json[path=.params.mapping, elide=12]`

4. It looks like this is a very similar requirement to the Mobile
  schema I'm working on, see e.g.
  https://docs.couchbase.com/sync-gateway/current/configuration-schema-database.html#_database

  (Takes a second to appear after loading, as it's generated
  dynamically)
  Notice how the key names are linked to the definitions below.

  Not quite identical use-case, but I wonder if this is something that
  we could look at generalizing?

osfameron committed:
Reindent JSON

sarahlwelton committed:
[AV-43732] Convert line includes to tags

osfameron commented on modules/search/pages/create-custom-token-filter.adoc:
```diff
@@ -42,7 +42,9 @@ You can create any of the following custom token filters:
-include::partial$custom-token-filters-descriptions.adoc[lines=2..3;9..27]
+include::partial$custom-token-filters-descriptions.adoc[tags=dict;!dict_example]
+
+// include::partial$custom-token-filters-descriptions.adoc[lines=2..3;9..27]
```
Probably tidier to remove these comments now.

osfameron reviewed (approved):
Thanks, looks good! 👍

sarahlwelton committed:
[AV-43732] Remove comments

sarahlwelton committed:
[AV-43732] Adding back in modules because I don't want to kill everything on merge
